apollo-router/src/query_planner/execution.rs

```diff
@@ -393,6 +400,58 @@ impl PlanNode {
             (value, errors)
         })
     }
+
+    /// This is a reimplementation of `ValueExt::deep_merge` which takes type hierarchies into account
+    /// when merging response objects. If we are merging two instances of `__typename`, we check the
+    /// schema to make sure we do not overwrite a subtype with an interface or parent type. Otherwise,
+    /// the data associated with the interface/parent type is removed from the response later on.
+    fn type_aware_deep_merge(schema: &Schema, target: &mut Value, source: Value) {
+        match (target, source) {
+            (Value::Object(a), Value::Object(b)) => {
+                for (key, value) in b.into_iter() {
+                    let k = key.clone();
+                    match a.entry(key) {
+                        Entry::Vacant(e) => {
+                            e.insert(value);
+                        }
+                        Entry::Occupied(e) => match (e.into_mut(), value) {
+                            (Value::String(type1), Value::String(type2))
+                                if k.as_str() == TYPENAME =>
+                            {
+                                // If type1 implements type2, or if it is a subtype of type2, we skip this overwrite
+                                // to preserve the more specific `__typename` in the response
+                                if !schema.is_implementation(type2.as_str(), type1.as_str())
+                                    && !schema.is_subtype(type2.as_str(), type1.as_str())
```

Comment:
You don't need the `is_subtype()` here, it's sufficient to just check `!schema.is_implementation(type2.as_str(), type1.as_str())`. That's because `__typename` can only ever be an object type (the common case), or an interface type (the `@interfaceObject` case). It can't be a union. It's also better for performance to skip the `is_subtype()` check.

Reply:
I thought that would be the case, but my simple test fails if I remove the `is_subtype` check.

Reply:
Ah, looking at the implementation of `Schema::is_implementation()`, it's just bugged. It's:
1. Fetching the given interface
2. Checking whether the interface implements the implementor (should be the other way around).

There's a `FIXME` there added by @SimonSapin , not sure why we didn't fix it.

For now at least, we can just use `!schema.is_subtype(type2.as_str(), type1.as_str())` and avoid the call to `Schema::is_implementation()`.